src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs

```diff
@@ -0,0 +1,361 @@
+
+        /// <summary>
+        /// Move the project tree down over one of its siblings.
+        /// </summary>
+        public static bool TryMoveDown(UnconfiguredProject unconfiguredProject, Project project, IProjectTree projectTree)
+        {
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(project, nameof(project));
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return TryMove(unconfiguredProject, project, projectTree, MoveAction.Down);
+        }
+
+        /// <summary>
+        /// Move the project tree down over one of its siblings.
+        /// </summary>
+        public static Task<bool> TryMoveDownAsync(ConfiguredProject configuredProject, IProjectTree projectTree)
+        {
+            Requires.NotNull(configuredProject, nameof(configuredProject));
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return TryMoveAsync(configuredProject, projectTree, MoveAction.Down);
+        }
+
+        /// <summary>
+        /// Determines if we are moving up or down files or folders.
+        /// </summary>
+        private enum MoveAction { Up=0, Down=1 }
+
+        /// <summary>
+        /// Creates a hash set lookup of project rooted file paths that are part of a given project tree.
+        /// </summary>
+        private static HashSet<string> CreateRootedPathLookup(UnconfiguredProject unconfiguredProject, Project project, IProjectTree projectTree)
+        {
+            var treeQueue = new Queue<IProjectTree>();
+            var lookup = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            treeQueue.Enqueue(projectTree);
+
+            // The queue is how we process each project tree.
+            while (treeQueue.Count > 0)
+            {
+                var tree = treeQueue.Dequeue();
+
+                if (tree.IsFolder)
+                {
+                    foreach (var childTree in tree.Children)
+                    {
+                        treeQueue.Enqueue(childTree);
+                    }
+                }
+                else
+                {
+                    if (tree is IProjectTree2 tree2 && IsValidDisplayOrder(tree2.DisplayOrder))
+                    {
+                        lookup.Add(unconfiguredProject.MakeRooted(tree.FilePath));
+                    }
+                }
+            }
+
+            return lookup;
+        }
+
+        /// <summary>
+        /// Get project item elements based on the project tree.
+        /// Project tree can be a folder or item.
+        /// </summary>
+        private static IEnumerable<ProjectItemElement> GetElements(UnconfiguredProject unconfiguredProject, Project project, IProjectTree projectTree)
+        {
+            var lookup = CreateRootedPathLookup(unconfiguredProject, project, projectTree);
+
+            var elements = new List<ProjectItemElement>();
+
+            foreach (var item in project.AllEvaluatedItems)
```

Comment:
Isn't this a pretty huge collection?  Can it be filtered to only `Compile` items or anything?

Reply:
It potentially can be, but I can't just make Compile items be the only thing we care about because we also order non-compilable items too.

In CPS though, they do filter by having an instance of `IProjectVersionedValue<IProjectItemSchema>` and getting a list of item types that we do care about: `var itemTypes = schemaData.Value.GetKnownItemTypes();` then they `project.AllEvaluatedItems.Where(item => itemTypes.Contains(item.ItemType))`.

Let me see what can be done here.

Reply:
I managed to find a much better way to handle this. The new approach simplified these methods as well.